examples/sod/ArborX_DetailsSOD.hpp

```diff
@@ -0,0 +1,388 @@
+      _critical_bin_values(pos).particle_index = getData(query);
+      _critical_bin_values(pos).halo_index = halo_index;
+      _critical_bin_values(pos).distance = dist;
+    }
+  }
+};
+
+// Compute radii for SOD bins
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<float **, MemorySpace>
+computeSODBinRadii(ExecutionSpace const &exec_space, float r_min,
+                   Kokkos::View<float *, MemorySpace> r_max, int num_sod_bins)
+{
+  Kokkos::Profiling::pushRegion("ArborX::SOD::compute_sod_bin_radii");
+
+  auto const num_halos = r_max.extent(0);
+
+  Kokkos::View<float **, MemorySpace> sod_halo_bin_outer_radii(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing,
+                         "ArborX::SOD::sod_halo_bin_outer_radii"),
+      num_halos, num_sod_bins);
+  Kokkos::parallel_for(
+      "ArborX::SOD::compute_bin_outer_radii",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_halos),
+      KOKKOS_LAMBDA(int halo_index) {
+        float r_delta = rDelta(r_min, r_max(halo_index), num_sod_bins);
+        sod_halo_bin_outer_radii(halo_index, 0) = r_min;
+        for (int bin_id = 1; bin_id < num_sod_bins; ++bin_id)
+          sod_halo_bin_outer_radii(halo_index, bin_id) =
+              pow(10.0, bin_id * r_delta) * r_min;
+      });
+
+  Kokkos::Profiling::popRegion();
+
+  return sod_halo_bin_outer_radii;
+}
+
+// Compute critical bins
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<int *, MemorySpace> computeSODCriticalBins(
+    ExecutionSpace const &exec_space, float rho, float rho_ratio,
+    Kokkos::View<double **, MemorySpace> sod_halo_bin_masses,
+    Kokkos::View<int **, MemorySpace> sod_halo_bin_counts,
+    Kokkos::View<float **, MemorySpace> sod_halo_bin_outer_radii)
+{
+  Kokkos::Profiling::pushRegion("ArborX::SOD::compute_critical_bins");
+
+  auto const num_halos = sod_halo_bin_masses.extent(0);
+  int const num_sod_bins = sod_halo_bin_masses.extent(1);
+
+  Kokkos::View<int *, MemorySpace> critical_bin_ids(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing,
+                         "ArborX::SOD::critical_bin_ids"),
+      num_halos);
+  Kokkos::deep_copy(critical_bin_ids, -1);
+  Kokkos::parallel_for(
+      "ArborX::SOD::compute_critical_bins",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_halos),
+      KOKKOS_LAMBDA(int halo_index) {
+        int &critical_bin_id = critical_bin_ids(halo_index);
+
+        double accumulated_mass = 0.;
+        for (int bin_id = 0; bin_id < num_sod_bins; ++bin_id)
+        {
+          accumulated_mass += sod_halo_bin_masses(halo_index, bin_id);
+          float outer_radius = sod_halo_bin_outer_radii(halo_index, bin_id);
+
+          float bin_rho_ratio_int = 0;
+          if (outer_radius > 0)
+          {
+            auto volume = 4.f / 3 * M_PI * pow(outer_radius, 3);
+            bin_rho_ratio_int = (accumulated_mass / volume) / rho;
+          }
+
+          if (bin_rho_ratio_int <= rho_ratio)
```

Comment:
Not sure. Lifted from CosmoTools. Maybe it's because of accuracy issues. The `rho` value is 1e14, and comparing large numbers is likely to be less accurate.